src/shortcuts/__tests__/extractThought.ts

```diff
@@ -0,0 +1,91 @@
+import { screen, findAllByText, findAllByPlaceholderText } from '@testing-library/react'
+import { extractThought, newThought } from '../../action-creators'
+import { store } from '../../store'
+import createTestApp, { cleanupTestApp } from '../../test-helpers/createRtlTestApp'
+import { setCursorFirstMatchActionCreator } from '../../test-helpers/setCursorFirstMatch'
+
+/**
+ * Set range selection.
+ */
+const setSelection = (element: HTMLElement, selectionStart: number, selectionEnd: number) => {
+  const range = document.createRange()
+  const sel = window.getSelection()
+
+  range.setStart(element.childNodes[0], selectionStart)
+  range.setEnd(element.childNodes[0], selectionEnd)
+
+  sel?.removeAllRanges()
+  sel?.addRange(range)
+
+  return range.toString()
+}
+
+describe('Extract thought', () => {
+  beforeEach(async () => {
+    await createTestApp()
+  })
+  afterEach(cleanupTestApp)
+
+  it('an alert should be shown if there is no selection', async () => {
+    store.dispatch([
+      newThought({ value: 'this is a thought' }),
+      newThought({ value: 'sub-thought', insertNewSubthought: true }),
+      setCursorFirstMatchActionCreator(['this is a thought'])
+    ])
+
+    store.dispatch([extractThought()])
+
+    const alert = await screen.findByText('No text selected to extract')
+    expect(alert).toBeTruthy()
+
+    const [thought] = await screen.findAllByText('this is a thought', { selector: 'div' })
+    expect(thought).toBeTruthy()
```

Comment:
We're getting closer. Extracting the first item in the array is not totally clear to me. Does `findAllByText` return more than one item? Why? What does `selector: `div` do in this case? 

The reason I am spending more time on this is that it still feels like there are implementation details mixed in with the test. If there are no alternatives then we can factor out the specifics into a custom `findThought` query. That will decouple the individual tests from the implementation details.

Reply:
Agreed. We need to abstract all the implementation details that we can, and I guess one way to do that is to create custom query utilities, as you mentioned.

Regarding your questions,

`findAllByText` returns an array by default as the name implies. We can use `findByText` to get rid of that.

The  `selector: div` is needed because there's no direct query to find only the visible elements, and since each thought comprises two elements - `subthought-text` (a hidden span) and an `editable` (the visible div), we need to be specific with our queries to get the correct(visible) element and avoid any array logic.

Reply:
> `findAllByText` returns an array by default as the name implies. We can use `findByText` to get rid of that.

Great! That's better.

> The `selector: div` is needed because there's no direct query to find only the visible elements, and since each thought comprises two elements - `subthought-text` (a hidden span) and an `editable` (the visible div), we need to be specific with our queries to get the correct(visible) element and avoid any array logic.

Ah, I see. Thank you for the explanation. I think selecting on something more semantic than the element type may be more transparent and understandable for other devs. It seems that the key difference is that the editable thought has a `content-editable` attribute. We should be able to create a custom query that is a variant of `findByText`. I would suggest calling `findByText` and then filtering its results by attribute rather than the other way around for efficiency. Let me know what you think of this approach.